### DISCUSSION TITLE
Python - FastAPI - Optional option for an UploadFile

### DISCUSSION THREAD

**Discussion by LidorPrototype** [fastapi/fastapi#8444] [Questions] [answered]
### First Check

- [x] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional
from fastapi import FastAPI, UploadFile, File

app = FastAPI()

@app.get("/")
def root():
    pass

@app.post("/OptionOne")
def pdf_foo1(file: Optional[UploadFile] = None):
    print(file.filename)

@app.post("/OptionTwo")
def pdf_foo2(file: Optional[UploadFile] = File(None)):
    print(file.filename)

@app.post("/OptionThree")
def pdf_foo3(file: UploadFile = File(None)):
    print(file.filename)

@app.post("/OptionFour")
def pdf_foo4(file: UploadFile = None):
    print(file.filename)

@app.post("/OptionFive")
def pdf_foo5(file: bytes = File(None)):
    print(file)
```

### Short Summary
I am trying to do what I think is a simple thing but no internet solution has helped me so far, and I didn't find anything helpful yet in the issues.
___
### The desired solution:
A route (function) that will accept an UploadFile but that will be optional, the user will not be have to provide it.

### Current Case:
In the code there are 5 different cases I tried, all of them are not working as follows:
Cases 1, 3 Gives an 'ValueError: Value not declarable with JSON Schema'
Cases 2, 4 Acts like it's not always must:
 - It will boot up and start working, but when I try to send a request (for example in the docs swagger) it will gives me a 422 error and will say it excepts data named "file" (in this case)
Case 5 Gives a different error (up until now I saw 3 different ones) but anyway using bytes is not the solution I'm looking for.

### Operating System

Linux, Windows

### Operating System Details

Im developing on a windows server, and publish on a Linux webapp (Azure)

### FastAPI Version

0.70.1

### Python Version

3.8.2

### Additional Context

I'm testing it with Swagger (Locally and Webapp) and also using the API links in a different program
( If I missed anything, or there is anything else you need to know in order to resolve this issue, please comment and I will update the data accordingly within 12 hours max [Not including Friday and Saturday] )

**Comment by acidjunk** [comment]
Does the example from the docs work?

```
from typing import Optional

from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/files/")
async def create_file(file: Optional[bytes] = File(None)):
    if not file:
        return {"message": "No file sent"}
    else:
        return {"file_size": len(file)}


@app.post("/uploadfile/")
async def create_upload_file(file: Optional[UploadFile] = None):
    if not file:
        return {"message": "No upload file sent"}
    else:
        return {"filename": file.filename}
```

https://fastapi.tiangolo.com/tutorial/request-files/#optional-file-upload

**Comment by acidjunk** [comment]
The upload assumes multipart form data: So when you see a JSON error; you might want to test your upload different.

**Comment by LidorPrototype** [comment]
Hey, if you will look at the options I wrote there, there is an option like that one in the docs (OptionOne and part of OptionFive but OptionFive is not good for me anyway)

**Comment by acidjunk** [comment]
Yes I saw that: no need to be rude.
Not sure how you test your upload. via SWAGGER? via a JS frontend?

I think I have used it like in the docs without probs. Used a more recent FastAPI though.

**Comment by LidorPrototype** [comment]
I'm sorry if it sounded rude, that was not my intention :)

I'm testing it with Swagger (Locally and Webapp) and using the API links in a different program ( I will update that info as well )

**Comment by acidjunk** [comment]
Can you confirm that the POST is not JSON?

'ValueError: Value not declarable with JSON Schema' seems to point to not correct HTTP content type.

**Comment by LidorPrototype** [comment]
Sorry for the delay, can you please walk me how to confirm it?
for the sake of my problem, Im doing the minimal, using the docs, trying to make a route that has for example 'Optional[UploadFile] = File(None)' or 'Optional[UploadFile] = None' and execute it via the docs but without uploading a file (can upload other variables, but not a must)

**Comment by acidjunk** [comment]
Open the browsers inspect element (often mapped to F12) than upload your file: you can see the POST and the headers in the network tab.

**Comment by LidorPrototype** [comment]
Hello again, yes I it is not a JSON, but I fail to see how it's related, b/c I want it to work even when I'm not sending a file (if I send a file it's working but if I'm trying to make it optional a.k.a without sending a file [sending None] it fails with the above errors

**Comment by jd-solanki** [comment]
I assume you want to process the file if uploaded and do nothing or skip the processing if file is not uploaded.

Please try below code:
```python
import aiofiles

from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/")
async def read_root(
    uploaded_file: UploadFile = File(None),
):
    r = {"file_status": "no file is provided"}
    if uploaded_file:
        async with aiofiles.open(f"{uploaded_file.filename}", "wb") as out_file:
            content = await uploaded_file.read()  # async read
            await out_file.write(content)

        r["file_status"] = "File is uploaded"
    return r
```

requirements.txt
```python
irements.txt 
aiofiles==0.8.0
anyio==3.5.0
asgiref==3.5.0
click==8.1.1
fastapi==0.75.0
h11==0.13.0
httptools==0.4.0
idna==3.3
pydantic==1.9.0
python-dotenv==0.20.0
python-multipart==0.0.5
PyYAML==6.0
six==1.16.0
sniffio==1.2.0
starlette==0.17.1
typing_extensions==4.1.1
uvicorn==0.17.6
uvloop==0.16.0
watchgod==0.8.1
websockets==10.2
```

In here if you upload the file then app will save it in your machine and if you don't upload a file then it will return "no file is provided".

**Comment by LidorPrototype** [comment]
Hello, I just tried your code, I still get a `422 Error: Unprocessable Entity` with the following response:
`{
  "detail": [
    {
      "loc": [
        "body",
        "uploaded_file"
      ],
      "msg": "Expected UploadFile, received: <class 'str'>",
      "type": "value_error"
    }
  ]
}`

**Reply by viniciusCalcantara** [reply]
The **"msg"** on the 422 Error says it's receiving a `str`, so just use:

`file: Annotated[UploadFile | str, File()] = None`

```python
from fastapi import FastAPI, UploadFile, File
from typing import Annotated

app = FastAPI()

@app.post("/")
async def read_root(
    file: Annotated[UploadFile | str, File()] = None
):
    if not isinstance(file, str):
        # file is an UploadFile
        ...
    else:
        # file is a string
        ...
```
I don't know if this solves your situation, but it worked for me.

**Comment by jd-solanki** [comment]
Which file you are uploading? I tried uploading
![wallpaer](https://user-images.githubusercontent.com/47495003/160991088-bf9813c5-bfb9-47ca-b528-68b2a609c45d.png)
above file via docs and didn't got any error.

**Comment by LidorPrototype** [comment]
I'm trying to make it work when I do not upload a file, when I upload a file it works fine, but when I don't upload a file I get the above error

**Comment by jd-solanki** [comment]
Have you tried with the above code in a separate env?

I get `no file is provided` if I don't upload the file.

Did you upload it via swagger docs?

**Comment by jd-solanki** [comment]
![chrome-capture-2022-2-31](https://user-images.githubusercontent.com/47495003/160992522-0af92272-a19f-4b8f-a434-d89f958ac233.gif)

**Comment by LidorPrototype** [comment]
@jd-0001 
I just tried your code with a 100% new project and env, and it does work...
do you have any idea why on my main project it does not work?

**Comment by jd-solanki** [comment]
Have you tested in your app via swagger docs?

Try adding this route as it is in your app and print the value of `uploaded_file`.

Edit: **Note that there's a checkbox saying `send empty value`.**

**Comment by LidorPrototype** [comment]
I tried your code in my main project, and when I'm sending it an empty file it gives the above error I mentioned, I looked at the request, it's the same post request but on my main project it wont work for some reason but on a brand new one it does, which is weird to me

**Comment by LidorPrototype** [comment]
The only difference I see is that in my main project the `upload_file `is showing without the "" besides it for some reason (in the request)
@jd-0001

**Comment by jd-solanki** [comment]
I am just confirming, have you checked `send empty value` in swagger?

Please try in incognito where every extension is disabled.

**Comment by LidorPrototype** [comment]
- Yes I checked the empty value box (the same as I did with your code in the new env)
- I tried incognito mode, still same problem
    - I'm in a computer without any internet extensions anyway

**Comment by jd-solanki** [comment]
The last thing I can tell is:

- check for middlewares

**If your project is big**
Try removing some other code from your app just for testing if you find it might cause the issue

**if your project is small**
Try to implement your project (if it's small) in this new example (just for finding the issue)

**Comment by jannik-el** [comment]
Does anyone have any suggestions how to create the same type of behavior for a multi file upload? 
Looks like this when it is not optional: 

```
@app.post("/uploadfiles/")
async def upload_files(subject_id: str, background_tasks: BackgroundTasks, files: list[UploadFile], ):
        # does something ...
        return {"success": "Files uploaded successfully."}
```
        
I need to make it optional. I have tried the following: 

```
# this one doesn't like it when you send an empty value since the value is a string
files: List[Union[UploadFile, None]] = File(None) 

 # this one does not allow the user to upload multiple documents
uploaded_file: UploadFile = File(None)

 # I can't upload docs through swagger-ui with this one which is needed, and does also not allow for uploading mutiple files
file: Optional[UploadFile] = None
```

I think part of the issue may also be caused by the background tasks - idk I'm out of my depth here, all comments, suggestions and help are much appreciated.

**Reply by JavierSanchezCastro** [reply]
try this:

`files: list[UploadFile] = []`

**Reply by jannik-el** [reply]
Unfortunately didn't quite work returned: 
Error 422 Unprocessable Entity

```
{
  "detail": [
    {
      "type": "value_error",
      "loc": [
        "body",
        "files",
        0
      ],
      "msg": "Value error, Expected UploadFile, received: <class 'str'>",
      "input": "",
      "ctx": {
        "error": {}
      },
      "url": "https://errors.pydantic.dev/2.5/v/value_error"
    }
  ]
}
```

This is when I return it with an empty value, not entering any value gives me: 
Error 400 
`
{
  "detail": "There was an error parsing the body"
}
`

**Reply by JavierSanchezCastro** [reply]
Yes sorry, you're right, forgot to say it only works with pydantic<2.

There's a PR to fix that https://github.com/tiangolo/fastapi/pull/9928

**Reply by viniciusCalcantara** [reply]
When we do the same as in  [FastAPI GitHub Discussions #8444](https://github.com/fastapi/fastapi/discussions/8444#discussioncomment-11598786)
(but not using Annotated this time)

We get:

```python
from fastapi import FastAPI, UploadFile
from typing import List

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

@app.post("/uploadfiles/")
async def create_upload_files(
    files: List[UploadFile] | str = None,
):
    if files == ['']:
        return {"filenames": []}
    
    return {"filenames": [file.filename for file in files]}
```

Then an error occurss:

```json
{
  "detail": [
    {
      "type": "value_error",
      "loc": [
        "body",
        "files",
        "list[function-plain[_validate()]]",
        0
      ],
      "msg": "Value error, Expected UploadFile, received: <class 'str'>",
      "input": "",
      "ctx": {
        "error": {}
      }
    },
    {
      "type": "string_type",
      "loc": [
        "body",
        "files",
        "str"
      ],
      "msg": "Input should be a valid string",
      "input": [
        ""
      ]
    }
  ]
}
```
we can see that it's waiting for a string but the input is a list of it.

so use: ```files: List[UploadFile] | List[str] = None```

```python
from fastapi import FastAPI, UploadFile
from typing import List

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

@app.post("/uploadfiles/")
async def create_upload_files(
    files: List[UploadFile] | List[str] = None,
):
    if files == ['']:
        return {"filenames": []}
    
    return {"filenames": [file.filename for file in files]}
```

and it works well.

"Note: The List[str] is used when an empty value is sent via the Swagger UI."

**Comment by Fabri2505** [comment]
Buenos días, sobre el problema anterior del envio de un archivo opcional a una api rest en FastApi. Me paso algo curioso. 
![image](https://github.com/user-attachments/assets/3ce30da6-1e13-42d2-be3f-279cbd509a30)

Solucione el problema desmarcando la opcion Send empty value. y le envie sin nada. Ahi si lo toma. Algo que me parecio extraño pero asi fue mi solucion. Espero que alguien pueda comprobar o refutar.

**Comment by YuriiMotov** [comment, marked as answer]
In current version (0.115.12) the solution will be:

```python
from typing import Annotated, Any

from fastapi import FastAPI, File, UploadFile
from pydantic import BeforeValidator
from pydantic.json_schema import SkipJsonSchema

app = FastAPI()

def empty_str_to_none(v: Any) -> Any:
    if v == "":
        return None
    return v

@app.post("/example")
def read_item(
    upload_file: Annotated[
        UploadFile | SkipJsonSchema[None],
        File(description="Some description"),
        BeforeValidator(empty_str_to_none),
    ] = None,
):
    print(upload_file)
```

https://github.com/fastapi/fastapi/discussions/10755#discussioncomment-13306945
